feat(lsp): Add LspAttach and LspDetach autocommands

gpanders opened the pull request:
The current approach of using `on_attach` callbacks for configuring buffers for LSP is suboptimal:

1. It does not use the standard Nvim interface for driving and hooking into events (i.e. autocommands)
2. There is no way for "third parties" (e.g. plugins) to hook into the event. This means that *all* buffer configuration must go into the user-supplied on_attach callback. This also makes it impossible for these configurations to be modular, since it all must happen in the same place.
3. There is currently no way to do something when a client detaches from a buffer (there is no `on_detach` callback).

The solution is to use the traditional method of event handling in Nvim: autocommands. When a LSP client is attached to a buffer, fire a `LspAttach` event with the client's name as the pattern. Likewise, when a client detaches from a buffer fire a `LspDetach` event.

This enables plugins to easily add LSP-specific configuration to buffers as well as enabling users to make their own configurations more modular (e.g. by creating multiple LspAttach autocommands that each do something unique).

gpanders committed:
feat(lsp): add LspAttach and LspDetach autocommands

The current approach of using `on_attach` callbacks for configuring
buffers for LSP is suboptimal:

1. It does not use the standard Nvim interface for driving and hooking
   into events (i.e. autocommands)
2. There is no way for "third parties" (e.g. plugins) to hook into the
   event. This means that *all* buffer configuration must go into the
   user-supplied on_attach callback. This also makes it impossible for
   these configurations to be modular, since it all must happen in the
   same place.
3. There is currently no way to do something when a client detaches from
   a buffer (there is no `on_detach` callback).

The solution is to use the traditional method of event handling in Nvim:
autocommands. When a LSP client is attached to a buffer, fire a
`LspAttach`. Likewise, when a client detaches from a buffer fire a
`LspDetach` event.

This enables plugins to easily add LSP-specific configuration to buffers
as well as enabling users to make their own configurations more modular
(e.g. by creating multiple LspAttach autocommands that each do
something unique).

ii14 commented on runtime/lua/vim/lsp.lua:
```diff
@@ -1348,6 +1349,8 @@ function lsp.buf_detach_client(bufnr, client_id)
   local namespace = vim.lsp.diagnostic.get_namespace(client_id)
   vim.diagnostic.reset(namespace, bufnr)
 
+  nvim_exec_autocmds("LspDetached", { pattern = client.name, modeline = false })
```
I think `LspDetached` also needs to be somewhere in `client.stop`?

gpanders replied:
Added it to the `on_exit` handler (which is called after `client.stop` so should handle this case too). Thanks.

clason commented:
Just to add my two (Euro)cents to it:

1. I strongly believe that an autocommand is indeed the best way of allowing third parties to hook into the LSP functionality. (Witness the monkey-patching around frameworks and LSP installer plugins.)
2. These autocommands _should_ be in core because otherwise you cannot expose the client (name) as `pattern` and have to rely on string manipulation.
3. This also makes it easier to, e.g., setup keymappings for _all_ LSP-attached buffers.

In fact, I think we should also add corresponding events for `Treesitter` (which would allow, e.g., emitting `syntax off` when treesitter is disabled -- which has been requested -- or setting `foldexpr` when using a language that supports folding).

Longer term, this would be a usecase for allowing user data in `v:event` (which here could be the actual client structure so we don't have to `get_client_by_name`).

justinmk commented:
https://github.com/neovim/neovim/pull/15949 added `User LspRequest`. If we are adding Lsp events to core (instead of `User`), it might make sense to promote `LspRequest` core, however.... 
let's first ask if it makes sense to have a more generic `ClientAttach` / `BufAttach` / ... event? 

> The new `vim.lsp.get_client_by_name()` function

What's holding us back from something like `vim.lsp.get_clients({predicate})` ?

clason commented:
> let's first ask if it makes sense to have a more generic ClientAttach / BufAttach / ... event?

It would, if we had a way of despatching on them based on _multiple_ patterns -- if we have to `:find` on a single string, we 
1. again have to rely on string matching, which I think is suboptimal;
2. are no better off than `User LspAttached-clangd`.

I also believe that a rule of thumb is that if you set up an autocommand for an event/pattern, you nearly always want to do something when it occurs. (I.e., an event/pattern that most of the time I would have to check something and return early is "wasteful" and not granular enough.)

> What's holding us back from something like vim.lsp.get_clients({predicate}) ?

That, on the other hand, sounds to me like a useful generalization. What other predicates could be useful here besides the client name? The buffer? @gpanders @mfussenegger

mfussenegger commented:
> What's holding us back from something like vim.lsp.get_clients({predicate}) ?

There is already `vim.lsp.get_active_clients()` and `vim.lsp.buf_get_clients(bufnr)` which one can use with `vim.tbl_filter(predicate, tbl)`. I don't really see the need to add another variant.


We could add a `tbl_find(predicate, table)` or a more generic `find(predicate, iterator_func|table)` (returning the first match or `nil`) for the use-case of finding a single client.

gpanders commented:
>>The new vim.lsp.get_client_by_name() function
>
>What's holding us back from something like vim.lsp.get_clients({predicate}) ?

As @mfussenegger said we already have the primitives available for this. `vim.lsp.get_client_by_name` is a convenience function as it is likely to be used often in these types of autocommands. The alternative is

```lua
local client = vim.tbl_filter(function(c) return c.name == args.match end, vim.lsp.buf_get_clients(args.buf))[1]
```

which is fine, but as soon as users need to write this more than once they are just going to make a `get_client_by_name` function anyway, so why not just provide it for them?

justinmk commented:
> have to rely on string matching, which I think is suboptimal;
> ...
> if we had a way of despatching on them based on multiple patterns

@clason good points, but "attach" is not a common event, even if we add "treesitter attach", "buffer attach", "LSP attach"--if many things are attaching, string-matching is a vanishingly-small part of the perf cost. So the interface matters more.

Do we want to add more core events for every "attach" kind? `BufAttach` has nice symmetry with `BufUnload`. Could also think about `WinAttach`, etc...

`v:event` can further disambiguate. 


> There is already `vim.lsp.get_active_clients()` and `vim.lsp.buf_get_clients(bufnr)`

Right, we want to avoid that continued pattern.

> We could add a tbl_find(predicate, table) or a more generic find(predicate, iterator_func|table) (returning the first match or nil) for the use-case of finding a single client.

Sounds great, but even before that, just to "stop the bleeding", we could generalize `get_clients()` (or whatever it's named if that already exists) as follows...

> they are just going to make a `get_client_by_name` function anyway, so why not just provide it for them?

If writing a predicate is too verbose (and if we can't fix _that_ yet), we can provide an interface like this:

```
get_clients(..., {name='foo'})
get_clients(..., {buf=42})
get_clients(..., {active=true})
```

Let's try to apply this idea everywhere. Including treesitter and whatever else. Let's stop adding zillions of special-purpose functions.

This pattern should be followed everywhere. We shouldn't have `get_thing_by_foo` in any Lua module.

Eventually, we'll have a super nice find() interface (maybe like @mfussenegger sketched out) so we won't even necessarily need the above `get_clients()` thing. https://github.com/neovim/neovim/issues/18585

justinmk commented:
>> BufAttach has nice symmetry with BufUnload. Could also think about WinAttach, etc... 
> but which is cheaper, all things considered: adding a new autocommand, or adding a new subsystem to filter?

Yeah, Buf/Win/... is probably not a useful discriminator for "attach" events. Do we think `ClientAttach [pattern=clientname]` also doesn't work?

So the only remaining request then is the `get_clients(..., <pseudofilter>)` idea.

gpanders committed:
feat(lsp): add filter to vim.lsp.get_active_clients()

Allow get_active_clients() to filter on client name, id, or buffer. This
(soft) deprecates lsp.buf_get_clients().

gpanders commented:
Done, I split the PR into two commits (one for the autocommands and one for `vim.lsp.get_clients`) so don't squash when merging.

clason commented:
That's a nice clean-up of the API! I suggest we mark the "Before" functions as deprecated in the docstrings (and maybe mention this in the "Following HEAD" issue), but wait with `vim.deprecate()` until after 0.8.0?

clason commented on src/nvim/auevents.lua:
```diff
@@ -70,6 +70,8 @@ return {
     'InsertEnter',            -- when entering Insert mode
     'InsertLeave',            -- just after leaving Insert mode
     'InsertLeavePre',         -- just before leaving Insert mode
+    'LspAttach',              -- an LSP client attached to a buffer
```
```suggestion
    'LspAttach',              -- after an LSP client attached to a buffer
```

clason commented on src/nvim/auevents.lua:
```diff
@@ -70,6 +70,8 @@ return {
     'InsertLeave',            -- just after leaving Insert mode
     'InsertLeavePre',         -- just before leaving Insert mode
+    'LspAttach',              -- an LSP client attached to a buffer
+    'LspDetach',              -- an LSP client detached from a buffer
```
```suggestion
    'LspDetach',              -- after an LSP client detached from a buffer
```

clason commented on runtime/doc/deprecated.txt:
```diff
@@ -106,11 +106,12 @@ internally and are no longer exposed as part of the API. Instead, use
 *vim.lsp.util.diagnostics_to_items()*	Use |vim.diagnostic.toqflist()| instead.
 *vim.lsp.util.set_qflist()*		Use |setqflist()| instead.
 *vim.lsp.util.set_loclist()*		Use |setloclist()| instead.
+*vim.lsp.buf_get_clients()*		Use |vim.lsp.get_active_clients()| instead.
```
```suggestion
*vim.lsp.buf_get_clients()*		Use |vim.lsp.get_active_clients({buffer = bufnr})| instead.
```

gpanders replied:
This will break the link though.

gpanders commented:
After discussion with @mfussenegger I've instead just modified `lsp.get_active_clients()` to take an (optional) filter table, which allows getting a client by name. This will also soft-deprecate `vim.lsp.buf_get_clients()`.

`get_client_by_id` is nice because it returns either a single client or `nil`, rather than a table. So it might make sense to keep this around in addition to `get_active_clients`.

clason commented on runtime/doc/lsp.txt:
```diff
@@ -461,6 +461,37 @@ LspSignatureActiveParameter
+    vim.api.nvim_create_autocmd("LspAttach", {
+      pattern = "*",
+      callback = function(args)
+        local client = vim.lsp.get_client_by_name(args.match)
```
`get_client_by_name` returns a table, right?

(also, need to fix the name when that is decided).

Probably want to filter on `{ buffer = args.buf, name = args.match}` here, too?

justinmk commented:
> get_client_by_id is nice because it returns either a single client or nil, rather than a table

We need to have a generalized solution for that (i.e. ask "why is it inconvenient to get a table?" and solve that). Not to have singular and plural forms of everything.

mfussenegger commented:
> > get_client_by_id is nice because it returns either a single client or nil, rather than a table
> 
> We need to have a generalized solution for that (i.e. ask "why is it inconvenient to get a table?" and solve that). Not to have singular and plural forms of everything.

It's not so much a problem of convenience, but of semantics. If a function knows that it either returns one result or none it can use an appropriate type and make that clear. I don't see how returning a list and promising that it will either contain zero or one elements is the better option. I don't think reducing API surface is warranted here.

clason commented:
Repeating here my alternative suggestion from chat: use the client id (which is unique) instead of the (freely chosen) server name as `pattern` instead. This will make it slightly more cumbersome to set up an autocommand for a specific server (such as `clangd`), but would not require additional API functions: e.g.,
```lua
nvim_create_autocmd('LspAttach',{
  callback = function(args)
    if vim.lsp.get_client_by_id(args.match).name == 'clangd' then
      <do something for clangd>
    end
  end})
```
As a bonus, you could handle multiple servers in a single autocommand by chaining `elseif`s.

gpanders commented:
Let's wait on https://github.com/neovim/neovim/pull/18613 (or something similar) which will allow us to pass client objects directly to the autocommand callback. Then we won't need to hijack the pattern to use the client name and can instead use the buffer name as intended.
